File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Guyana.m
//
//  UIImage+Guyana.m
//  CountriesFlags
//

#import "Countries/UIImage+Guyana.h"

#import "Extensions/UIBezierPath+Triangle.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (Guyana)

+ (UIImage * _Nullable)flagGuyanaWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:239.0f/255.0f
                                        green:51.0f/255.0f
                                         blue:64.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:209.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:151.0f/255.0f
                                           blue:57.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorBlack = [UIColor blackColor];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    [[UIBezierPath triangleInSize:CGSizeMake(size.width * 2.0f, size.height)
                             type:FlagTriangleTypeLeftMiddle] fill];
    
    [colorYellow setFill];
    UIBezierPath* path = [UIBezierPath triangleInSize:CGSizeMake(size.width / 1920.0f * 3570.0f, size.height / 1152.0f * 1078.0f)
                                                 type:FlagTriangleTypeLeftMiddle];
    [path translateToPoint:CGPointMake(0.0f, size.height / 1152.0f * 37.0f)];
    [path fill];

    [colorBlack setFill];
    [[UIBezierPath triangleInSize:size type:FlagTriangleTypeLeftMiddle] fill];

    [colorRed setFill];
    path = [UIBezierPath triangleInSize:CGSizeMake(size.width / 1920.0f * 1772.0f, size.height / 1152.0f * 1062.0f)
                                   type:FlagTriangleTypeLeftMiddle];
    [path translateToPoint:CGPointMake(0.0f, size.height / 1152.0f * 43.0f)];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Guyana.swift
//
//  UIImage+Guyana.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Guyana "wikipedia"
     
     Generates the flag of Guyana.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Guyana.
     */
    internal static func flagGuyana(with size: CGSize) -> UIImage {
        let colorRed = UIColor(red: 239.0 / 255.0,
                               green: 51.0 / 255.0,
                               blue: 64.0 / 255.0,
                               alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 209.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)
        let colorGreen = UIColor(red: 0.0,
                                 green: 151.0 / 255.0,
                                 blue: 57.0 / 255.0,
                                 alpha: 1.0)
        let colorBlack = UIColor.black
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        UIBezierPath.triangle(in: CGSize(width: size.width * 2.0, height: size.height),
                              type: .leftMiddle).fill()

        colorYellow.setFill()
        var triangleSize = CGSize(width: size.width / 1920.0 * 3570.0, height: size.height / 1152.0 * 1078.0)
        var path = UIBezierPath.triangle(in: triangleSize, type: .leftMiddle)
        path.translate(to: CGPoint(x: 0.0, y: size.height / 1152.0 * 37.0))
        path.fill()

        colorBlack.setFill()
        UIBezierPath.triangle(in: size, type: .leftMiddle).fill()

        colorRed.setFill()
        triangleSize = CGSize(width: size.width / 1920.0 * 1772.0, height: size.height / 1152.0 * 1062.0)
        path = UIBezierPath.triangle(in: triangleSize, type: .leftMiddle)
        path.translate(to: CGPoint(x: 0.0, y: size.height / 1152.0 * 43.0))
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
